### ISSUE TITLE
ANTLR plugin doesn't refresh tree when modifying included grammar

### ISSUE THREAD

**Issue by wstrzalka** [antlr/intellij-plugin-v4#242]
I have grammar & lexer files split.
When I do modify grammar the ANTLR Preview refreshes the view but when I modify grammar it doesn't.
I include the grammar via tokenVocab option.

Plugin version 1.8

**Comment by parrt** [comment]
tokenVocab doesn't include; it just gets token types. Are the files named XParser.g4 and XLexer.g4? If not, that's a problem for the plugin.

**Comment by melihovv** [comment]
I have the same issue. Files are named LemonParser.g4 and LemonLexer.g4.

**Comment by lauriecharlwood** [comment]
I'm running Webstorm 2016.3 and have the same issue. Trying to use the mode function of the lever so need the separate files. Wondering if it's an IDE issue rather than a plugin issue.

**Comment by parrt** [comment]
@lauriecharlwood @melihovv @wstrzalka are you hitting "save"? It appears to work for me when I save. I am changing a lexer rule, saving, flippling back to parser grammar and its tokens in preview window are correctly relexed.

**Comment by shadrina** [comment]
Unfortunately, I faced the same problem. Saving doesn't help, so sometimes I have to restart the IDE after changing something in a lexer grammar.

**Comment by parrt** [comment]
Can someone post a sample split lexer/parser grammar so I can test? It works for me.

**Comment by AlexCouch** [comment]
@parrt I'm providing some examples for you to use. There might be problems with my grammar but for some reason, the parse tree will only show the first line parsed, same with hierarchy view. I can open up a separate issue for this problem but I am also have the same problem where every time I make any change, I have to right click on a rule and test the rule again. Literally every time. I'm on a mac os X sierra and I'm running in IntelliJ IDEA 2017.2.5. Currently trying to test my variable statement. It works but I don't know if it works for more than one line cause it doesn't do anything no matter what I do.

Here is my lexer: https://gist.github.com/AlexCouch/1ef78fb216b939d23a2a780431db7403
Here is my parser: https://gist.github.com/AlexCouch/754eb00353761f4a3224d8091a38d818

**Comment by Tegawende** [comment]
I have the same problem and i have to restart Intellij after i made some changes in the lexer.
Otherwise ANTLR Preview doesn't take into account my changes.

**Comment by bjansen** [comment]
I found a case where I modify the lexer grammar, hit save and get strange results in the ANTLR preview window because the `.tokens` file was not updated, so the parser grammar still uses the old `.tokens` file.

Here's my scenario:
* import both grammars mentioned in @AlexCouch's comment
* test rule `block` with the following input:
```
{
    if (1 foo == 2) {}
}
```
which gives this error:
> line 2:10 mismatched input 'foo' expecting VarCall

* change the token definition of `VarCall` like this:
```
VarCall             :   'bastien'                                           ;
```
* hit save and come back to the parser grammar, the preview is correctly updated:
> line 2:10 mismatched input 'foo' expecting 'bastien'

* revert the modification done in the lexer grammar, hit save and come back to the parser grammar, the preview is *not* updated:
> line 2:10 mismatched input 'foo' expecting 'bastien'

I think I found how to fix the problem, I just need to test it a bit more. I hope that I ran into the same problem as other people have here.

BTW having to explicitly save files is not something that IntelliJ users are expected to do, so it is not very intuitive to save files to test changes in lexer grammars. I suppose we could improve the plugin and make any save action unnecessary.

**Comment by bjansen** [comment]
All right, could you guys please test this build of the plugin and let me know if it fixes your problems?

[intellij-plugin-v4-1.9.zip](https://github.com/antlr/intellij-plugin-v4/files/2928161/intellij-plugin-v4-1.9.zip)

Download the zip then follow the procedure named ‘Install plugin from disk’ on [this page](https://www.jetbrains.com/help/idea/managing-plugins.html). ATM you will still need to hit save before switching back to your parser grammar.

**Comment by Tegawende** [comment]
It still have the same problem.

`grammar B314;
import B314Words;
root: ID;`

`lexer grammar B314Words;

// Identifiers
ID: LETTER (LETTER | DIGIT)* ;
NUMBER: (DIGIT)+;
fragment LETTER: 'A'..'Z' | 'a'..'z' ;
fragment DIGIT: '0'..'9' ;

// Comments -> ignored
//COMMENT: '/*' .*? '*/' -> skip;

// Whitespaces -> ignored
NEWLINE: '\r'? '\n'  -> skip ;
WS: [ \t]+ -> skip ;`

I used the parser and the lexer above.
When I put the COMMENT part in comment, even after I save the lexer file, ANTRL preview keep letting me using COMMENT  as if the lexer has not been modified.
Restart Intellij still fixe the problem.

**Comment by bjansen** [comment]
OK, thanks for the detailed use case, I will look into it.

**Comment by bjansen** [comment]
@parrt does [this fix](https://github.com/antlr/intellij-plugin-v4/commit/e12d8b8d2dcb059037bd62aba008a880be4e115a) look correct to you? 

@Tegawende is `import`ing his lexer grammar instead of using `tokenVocab`. This plugin didn't refresh the importing grammar in that case. I also added a TODO because I believe there could be more than one `PreviewState` to refresh, if multiple grammars use import/tokenVocab for this lexer grammar. Do you agree with me?

**Comment by bjansen** [comment]
@Tegawende in case you want to test a second fix: [intellij-plugin-v4-1.9.zip](https://github.com/antlr/intellij-plugin-v4/files/2932866/intellij-plugin-v4-1.9.zip)

**Comment by Tegawende** [comment]
Same problem, ANTLR preview is not correclty updated after I modify and save the lexer.  

However I noticed that ANTLR preview is correclty updated if the parser file is closed when I modify the lexer or if I close and reopen the parser file after I modified the lexer. 

It's already much better than restart intellij.

**Comment by bjansen** [comment]
Do you have other parser grammars that import this same lexer grammar by any chance?

**Comment by Tegawende** [comment]
No, i just have one parser.

**Comment by parrt** [comment]
@bjansen nothing scary from quick check of the code. thanks!

**Comment by bjansen** [comment]
@Tegawende I can't reproduce this specific problem anymore, with the fix I provided above:

> When I put the COMMENT part in comment, even after I save the lexer file, ANTRL preview keep letting me using COMMENT as if the lexer has not been modified.

Do you still have this problem, or is it some other problem you have now?

**Comment by blueglyph** [comment]
I'm not sure this is the same issue but it seems so.

With ANTLR v4 grammar plugin v1.9, in IntelliJ IDEA CE 2019.1.1 (so the up-to-date versions):

I have a lexer and grammar in two separate files, and I'm using the ANTLR Preview to construct both at the moment.

Problem:

Once the preview is open, after 1 or a few modifications of the lexer, the preview doesn't take the new versions into account and keeps behaving as if there were no changes.
Also, clicking on the close button (cross) which is just to the right of the "ANTLR Preview" panel title empties the panel content, doesn't close the panel, and makes it impossible to use anymore.
In either situation, it seems a combination of restarting IntelliJ IDEA / clearing the generated code is necessary.

Only modifying the parser seems OK, but I think I had the same problem after a number of modifications, I'm just not sure about that part. The lexer modification is easily reproducible (and is actually making the Preview pretty useless or very frustrating to use).

Is there a proper way to close this preview so it can be restarted after modification of the lexer source code? Is it necessary to manually clean some of the files? I assume it's not and the problem described above by others.

**Comment by blueglyph** [comment]
Is the fix mentioned earlier already in the "official" 1.9? If not, I can try and post the lexer/grammar here if that helps, since I have the same problem right now.

I'm using IntelliJ IDEA 2019.1.1 on Win7 x64, and the plugin version 1.9 (installed normally through IntelliJ). The ANTLR4 libraries are version 4.7.2 (though I don't think this is relevant here).
There is one parser, and the lexer / grammar in two separate files, they generate Java source code.

Modifying the lexer only once or a very few times, after the ANTLR Preview is started, usually fails to bring any update to the preview. For the grammar, failures are less frequent but seems to happen ultimately after repeated modifications (I didn't try to get a systematic, reproducible symptom).

Since the 'auto-generate on save' doesn't seem to trigger every time, I re-generate the lexer then the grammar manually every time.

I'm not sure whether this is linked or not, but there doesn't seem to be a proper way to close the Preview, like other Debug/Console/... tools. There is a cross just to the right of the "ANTLR Preview" title but clicking on it only puts the Preview in a broken state and IntelliJ must be restarted.

**Comment by bjansen** [comment]
@blueglyph no, it's not in 1.9, but you can try the fix in the link I provided: https://github.com/antlr/intellij-plugin-v4/issues/242#issuecomment-469861357

If you still have the same problem with this build, it would indeed be great if you could post your grammars, and the steps that can be followed to reproduce your problem.

**Comment by blueglyph** [comment]
@bjansen I did give it a try but I see no difference.

To be clear, I installed the plugin from disk, saw that the icon changed. I made a project from scratch by extracting a part of the lexer/parser, did the test after resetting the IDE, and it's reproducible.

The .g4 files are in <project>/lang

Below is the procedure I followed, let me know if I did something wrong on my side, or if you can reproduce the problem. I put 3 alternatives to try and get the expected result, the only that succeeds is by unloading the lexer & parser files from the editor and loading them again.

Of course, doing a restart of the IDE also works, I didn't explicitly mention that in the procedure below.

Thanks!

[lexer_test.zip](https://github.com/antlr/intellij-plugin-v4/files/3145795/lexer_test.zip)

- save lexer
- generate lexer
- save parser
- generate parser
- open TestParser.g4
- right-click on dbg_calc / Test rule dbg_calc
- enter test pattern:
"0x20;<return>" (zero - letter x - two zero) -> OK
"0o02;<return>" (zero - letter O - zero two) -> line 2:1 extraneous input 'o02' expecting ';'(this is normal)

- uncomment octal-related parts lines 130-139 and 146
- save lexer
- generate lexer
- save parser
- generate parser
- select TestParser.g4

-> the patterns are still there, but the error remains

- click on ANTLR Preview (to close the panel)
- right-click on dbg_calc / Test rule dbg_calc

-> same outcome (the pattern is still there)

(point 1)

--------------------------------------

(a) from point 1, tentative 1, closing the parser file

- close TestParser.g4
- open TestParser.g4
- click on ANTLR Preview (to close the panel)
- right-click on dbg_calc / Test rule dbg_calc
- enter test pattern:
"0o02;<return>" (zero - letter O - zero two) 

-> line 2:1 extraneous input 'o02' expecting ';'(this is normal)
-> same outcome

--------------------------------------

(b) from point 1, tentative 2, closing parser and lexer files

- close TestParser.g4, TestLexer.g4
- open TestParser.g4, TestLexer.g4
- click on ANTLR Preview (to close the panel)
- right-click on dbg_calc / Test rule dbg_calc
- enter test pattern:
"0o02;<return>" (zero - letter O - zero two) 

-> OK

--------------------------------------

(c) from point 1, only closing the panel

- click on the cross right of "ANTLR Preview" panel title

-> panel is empty, it is not possible to get the preview anymore

**Comment by blueglyph** [comment]
Maybe it was too long, let me know if I can help narrow down the problem.

**Comment by bjansen** [comment]
@blueglyph sorry I haven't had time to look at the problem yet, but I believe your detailed scenario will help me pinpoint the problem.

**Comment by bjansen** [comment]
OK, your comment was very helpful and I am able to reproduce your problem. I found a workaround that does not involve closing files or restarting the IDE :
* modify the lexer grammar, then save
* switch to the parser grammar, make a random modification like adding a blank line, then save

However this feels *very* hacky. I believe there shouldn't be any need to save files or generate code. Simply modifying the lexer and switching back to the parser should be enough to update the preview automatically. This will definitely improve the UX, so I'll start working on that.

**Comment by blueglyph** [comment]
Thanks for the feedback and the work-around!

**Comment by bjansen** [comment]
Nope, that commit was for #360 :/

**Comment by bjansen** [comment]
This build should hopefully fix the problem: [intellij-plugin-v4-1.10.zip](https://github.com/antlr/intellij-plugin-v4/files/3359129/intellij-plugin-v4-1.10.zip)
